### ISSUE TITLE
[Withdrawn] SDL 0270 - Add means to launch with optimum appHMIType

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#892]
Hello SDL community,

The review of the revised proposal "SDL 0270 - Add means to launch with optimum appHMIType"  begins now and runs through February 18, 2020.  The original review of this proposal took place December 11 - December 17, 2019.  The proposal is available here:

https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0270-Add-means-to-launch-with-optimum-appHMIType.md

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/892

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by joeljfischer** [comment]
> The operation after SDL App receives launchAppHMIType is not specified.

It's difficult to understand the purpose of this proposal if the author is not going to say what an app should do about it. I believe that this proposal needs several examples of apps that register with multiple `appHMIType`s and receive this information and how they alter their operation based on it in order to understand the usefulness of this proposal.

**Comment by Shohei-Kawano** [comment]
Hi @joeljfischer -san

Thank you for your comment.
And, I'm sorry for I didn't write a concrete example in the proposal docs.

As an example in Japan, On TOYOTA IVI devices, MediaType's SDL App can be called with an icon touch from both the Audio function and the connectivity function.
LINE's "LINEカーナビ" is based on a navigation system, but user can use the Clova (LINE's VR service) function to play LINE music services on the "LINEカーナビ".

Currently, when "LINEカーナビ" is lunched with TOYOTA IVI,
Although the music service can be used with "LINEカーナビ", the icon does not appear on the Audio function screen.
Even if the icon is displayed, the "LINEカーナビ" cannot know that it has been selected from Audio, so it will be no different from when it is called from the Connectivity function.

When I looked at this situation, I thought that this proposal could provide a good UX, so I proposed this.

* "LINEカーナビ" is LINE's Navigation App.

**Comment by joeygrover** [comment]
I think an alternative to this proposal would be no changes to the project, but the apps using two different `appHMITypes` create two different SDL implementations. The two implementations would send individual RegisterAppInterfaces with specific `appHMITypes` and then it is clear when each is launched what they should do. This can be accomplished by create two services for Android for example. I understand this will not work for all cases, but it definitely should be considered an alternative and be compared to the cost of adding such a parameter and the effort required by developers to handle multi-state apps.

1. Would this `launchAppHMIType` be sent for every `OnHMIStatus `? Or just `HMI_FULL` status?
2. Maybe a better naming for this is `appHMITypeContext` to describe the context of this particular `OnHMIStatus`? 
3. I think we would need additional documentation and flow charts on how this should be used before implementation as it adds yet another state to HMI status.
4. What would happen in the case the app was first launched in Navigation mode, starting the video streaming manager classes, then the app was launched in the MEDIA mode after?  The current video streaming managers would automatically receive the `OnHMIStatus` and atttempt to start streaming again. I believe the impact to the manager layers must be considered here and include how they should operate.

**Comment by joeljfischer** [comment]
> I think an alternative to this proposal would be no changes to the project, but the apps using two different appHMITypes create two different SDL implementations.

This is currently not possible using the iOS library at least, but I do agree that it would be a better solution. That way LINE would have "LINE (Music)" and "LINE (Navigation)" separate apps appear on the screen even if they come from the same SDL implementation. However, I do believe it would be a huge amount of work on the iOS side to make this work.

> What would happen in the case the app was first launched in Navigation mode, starting the video streaming manager classes, then the app was launched in the MEDIA mode after?

This is the difficult and bad case. Would the `launchAppHMIType` have two different HMI levels? Would it tell the app to switch and stop navigation and start media streaming? Would the app need to send `SetDisplayLayout`? This proposal needs a lot more information to describe what is intended and what would happen in every case.

**Comment by theresalech** [comment]
The Steering Committee voted to return this proposal for revisions.  The author will add use case examples to the proposal.  The author will also investigate alternative options per the Project Maintainer’s recommendations, and provide their findings in the proposal – either as a new solution, or within the “Alternatives Considered” section, pending the results of their investigation.

**Comment by joeygrover** [comment]
I don't believe the previously stated questions were answered by the updates to this proposal. The author has added a few example cases but has otherwise not satisfied the other items. 



These still need to be answered:

1. Would this `launchAppHMIType` be sent for every `OnHMIStatus `? Or just `HMI_FULL` status?
2. Maybe a better naming for this is `appHMITypeContext` to describe the context of this particular `OnHMIStatus`? 
3. I think we would need additional documentation and flow charts on how this should be used before implementation as it adds yet another state to HMI status.
4. What would happen in the case the app was first launched in Navigation mode, starting the video streaming manager classes, then the app was launched in the MEDIA mode after?  The current video streaming managers would automatically receive the `OnHMIStatus` and atttempt to start streaming again. I believe the impact to the manager layers must be considered here and include how they should operate.

Additionally,
5\. The author added that the integration burden would be increased by using multiple specific SDL instances (multiple services on Android) but does not specify how. It is also not stated that this proposed solution would also increase the integration burden. The app will have to maintain `n` separate HMI statuses (where `n` is the number of hmi types the application supports). This also relates to number 3 and 4.

**Comment by Kazuki-Sugimoto** [comment]
@joeygrover 
Thank you for your comment.

> 1. Would this launchAppHMIType be sent for every OnHMIStatus ? Or just HMI_FULL status?

Basically, it is my recognition that it is sent only when HMI_FULL, but it is also necessary to consider cases such as LIMITED.

> 2. Maybe a better naming for this is appHMITypeContext to describe the context of this particular OnHMIStatus?

Yes. I think it is a good idea to change the name from launchAppHMIType to appHMITypeContext.

> 3. I think we would need additional documentation and flow charts on how this should be used before implementation as it adds yet another state to HMI status.

Regarding additional documentation and flow charts, I will add it.

> 4. What would happen in the case the app was first launched in Navigation mode, starting the video streaming manager classes, then the app was launched in the MEDIA mode after? The current video streaming managers would automatically receive the OnHMIStatus and atttempt to start streaming again. I believe the impact to the manager layers must be considered here and include how they should operate.

I think there are two cases.
Case1. Change from Projection screen to Template screen
Case2. Play music on the back of the Projection screen
In case 1, the app does not resume streaming, and in case 2, it needs to resume streaming. 
I think it is necessary to add a mechanism that allows the app developer to select it as an option.

> 5. The author added that the integration burden would be increased by using multiple specific SDL instances (multiple services on Android) but does not specify how. It is also not stated that this proposed solution would also increase the integration burden. The app will have to maintain n separate HMI statuses (where n is the number of hmi types the application supports). This also relates to number 3 and 4.

Do you want the details of the alternatives?
If so, I will add the description.

**Comment by joeljfischer** [comment]
4\. Regarding this point, could you clarify the following statement:

> I think it is necessary to add a mechanism that allows the app developer to select it as an option.

I'm not sure what is being noted here.

I'm also not sure if that's what @joeygrover was asking here. The video streaming managers would need to be altered to not start re-start video streaming.

Additionally, when this feature is available on a head unit, any developer that doesn't properly handle this by changing their template will cause extremely poor UX on the head unit (e.g. the user selects an app using the nav system and the app appears with their media template). I can't currently think of any mitigation if the developer doesn't manually update their app to support this.

6\. Are `NAVIGATION` and `MEDIA` the only supported use-cases right now?

**Comment by Kazuki-Sugimoto** [comment]
@joeljfischer -san
>The video streaming managers would need to be altered to not start re-start video streaming.

I understood not to restart video streaming. The default is to do so.
However, for scalability, I suggest that the app can select also not to stop video streaming.

>Additionally, when this feature is available on a head unit, any developer that doesn't properly handle this by changing their template will cause extremely poor UX on the head unit (e.g. the user selects an app using the nav system and the app appears with their media template). I can't currently think of any mitigation if the developer doesn't manually update their app to support this.

It's up to the app to set up multiple appHMITypes, so the app need to guarantee UX.

>Are NAVIGATION and MEDIA the only supported use-cases right now?

NAVIGATION and MEDIA are examples. Other types of combinations will be supported.

**Comment by joeljfischer** [comment]
@Kazuki-Sugimoto Because you state that the app needs to guarantee the UX, I must ask that you list every possible supported combination and their behavior. In addition, the proposal will need to add that information to documentation to ensure that every app knows what to do, and we will have to add that information to the app testing plan to ensure that every app does it properly.

**Comment by joeygrover** [comment]
5\. I've thought through this a few times and honestly I have to oppose this proposal. The amount of overhead this creates in the library (specifically the manager layer) is just too much to be worth the effort. I would much rather see the library be extended to support multiple SDL app registrations from a single mobile app that than a single SDL app try to create multiple app experiences. The feature as described in the proposal would be limited to only be functional with IVI systems that hardcoded app buttons/icons into their "connected app" screen. That means OEMs would have to know which apps support multiple HMI types, what those HMI types are and what to display for each of those HMI types. This creates an unnecessary burden on OEMs to maintain such an integration. While the author stated that the "two services" method would increase the development burden, I disagree. It is much cleaner to create services or classes that are scoped to perform specific functions and be controller to specific functionality that it is to mesh them all together in a monolithic integration. 

The current SDL app library stack is setup for a single, SDL app type integration. I would vote to create a proposal to allow mobile apps to register multiple SDL apps through the library. This would allow us to branch the stack where it makes sense instead of rooting this feature through the entire stack and creating a feature that only a few apps would use.

**Comment by Shohei-Kawano** [comment]
@joeljfischer -san, thank you for your comment.

5. I don't think our proposal is a permanent good idea. There are actually cases like LINE Car Navigation App and HU of TOYOTA, so we propose them as a way to improve UX at low cost.
I also think that the permanent good idea is the one Joel is talking about.
In the Japanese market, navigation is very basic and always used. Many users use music content while displaying navigation.
Therefore, there is a need for a mechanism that allows multiple services and apps to operate in the short and medium to long term.
The short-term method is what we suggested. I think the medium- to long-term method will be the one proposed by Joel.
If what we suggest is not a good short-term solution, do you have any good other ideas?

**Comment by joeljfischer** [comment]
Unfortunately, I think that we cannot do a short-term option and a medium / long-term option. That merely means that we would have two supported ways to do the same functionality forever, which isn't maintainable. I don't think that we can or should say "we know this isn't the right way to do it, but let's implement it for now."

Additionally, because this proposal wouldn't be implemented until Fall 2020 at the earliest, _both_ your solution and Joey's are what I would say are "long-term" solutions. Since that's the case, we should only take the right path, not both the wrong and right paths.

**Comment by Shohei-Kawano** [comment]
@theresalech -san,

First of all, I have an opinion on "we know this isn't the right way to do it, but let's implement it for now.".
I do not consider this proposal the wrong way.
When thinking permanent, I think my proposal is not enough because the scope is narrow,
However, in the short term, I think my proposal is right and made suggestions.

If Joey-san's solution is already implemented,
I will make proposals that extend it.
However, you said that it is not implemented at this time and will take time to implement.

Actually, LINE Car Navigation is released through App certification, and SDL-compatible HUs are installed on Toyota vehicles.
If Joey-san's solution is different from the LINE's thought, provide an extended mechanism,
or I think it is necessary to take care so that the thoughts are aligned.

How do you think it is right to handle such a case?

**Comment by joeljfischer** [comment]
I believe that this was directed at myself and Joey, and not Theresa, so I will respond.

As Joey and I have noted, there are several large downsides to your proposal.
1. Your proposal does not fully enumerate the various possible states as I requested above, given that you said that many other combinations of `appHMIType`s are possible. This will greatly increase the complexity of this feature.
2. Given that we both seem to agree that Joey's solution is better than this one, implementing this one would give us two solutions for the same problem, one better, and one worse, that now all HMI integrators and/or app developers must integrate both. This is an increase in maintenance work and integration work for everybody, and it is confusing.
3. Based on the current proposal, the app developer cannot turn this on or off, which means that any app that has the _capabilities_ for this feature (i.e. they use the updated library and have both navigation and music capabilities, or any of the others you mentioned were possible but didn't list) but does not integrate it properly will be an unavoidably poor UX on any head unit integrating this feature.

I'm confused by your note that LINE is released and Toyota has SDL vehicles on road. Is this intended to be a point to show that your current proposal is a short-term solution? If so, then your point doesn't follow from the statement. For this feature to be implemented, it would need to be developed and implemented into the next release of Core (earliest Fall 2020), then developed into the Toyota HMI, then updated into the vehicles on the road, and the updated library would need to be integrated into LINE, and they would have to update to support it. That means that this "short-term" solution would take years.

On the other hand, if Joey's solution were attempted and a single app could register multiple SDL apps, that would be an app-library only change, _not_ a Core change, which means that it once it were implemented into the app library and into LINE, all current Toyota vehicles should automatically support it. This is a large app-library change, but the net result should be that it could be integrated much quicker than your current solution.

In other words, we both seem to agree that Joey's solution is the better one, and my point is that his solution is also shorter-term compared to this one.

**Comment by theresalech** [comment]
During the 2020-02-18 meeting, the Steering Committee voted to keep this proposal in review, to allow for further discussion on this review issue.

**Comment by Shohei-Kawano** [comment]
@joeljfischer -san
I'm sorry, my mention was wrong...

I understand your explanation. It's certainly not a good idea to implement several similar solutions at the same time.

By the way, the LINE app has multiple services. I think that it has Navigation, Media and VoiceAssistant.
Is it a match from the  AppService?
If not, I think the maintainer needs to support.

**Comment by joeljfischer** [comment]
The suggestions here aren't related to app services. I will quote @joeygrover's comment:

> The current SDL app library stack is setup for a single, SDL app type integration. I would vote to create a proposal to allow mobile apps to register multiple SDL apps through the library. This would allow us to branch the stack where it makes sense instead of rooting this feature through the entire stack and creating a feature that only a few apps would use.

The change would be to allow an app to register multiple applications and does not need app services. 

However, to respond to your comment about app services, there is no current Voice Assistant app service. There have been proposals in the past, but none have been adequate.

**Comment by theresalech** [comment]
The Steering Committee voted to keep this proposal in review to allow the author time to respond to the comments on the review issue.  It was noted that it's very important the author respond to the issues in a timely manner, as we'll need to make a decision on this proposal during next week's call.

**Comment by Shohei-Kawano** [comment]
@joeljfischer -san
Sorry for late reply... I was having a little trouble...

Let me know about your ideas.
Is it like creating multiple SDLManagers, assigning each service to each and running them simultaneously?

**Comment by joeljfischer** [comment]
I think those ideas could be hashed out in a separate proposal, but generally yes, the app would have multiple `SDLManager`s running simultaneously.

**Comment by Shohei-Kawano** [comment]
@joeljfischer -san
Thanks comment.

In other words, it is good to propose a solution that allows multiple services to run on App.
And my proposal is part of this feature. Is my understanding right?
You mean that we need to make extensive proposal, right?

**Comment by joeljfischer** [comment]
Hi @Shohei-Kawano, I think that the current proposal should be withdrawn or rejected because it is not the best way to solve the problem that you are trying to solve. I don't believe that any part of this current proposal is needed. Instead, a new proposal needs to be made to support multiple `SDLManager`s running simultaneously on iOS, Android, and maybe JavaScript libraries. There seems to be an [in-progress PR for this](https://github.com/smartdevicelink/sdl_evolution/pull/922) but it has not been updated in a while.

**Comment by Shohei-Kawano** [comment]
@joeljfischer -san
Thank you for valuable information.

>There seems to be an in-progress PR for this but it has not been updated in a while.
Do you know the link or # information of the PR?
We want to refer to it, analyze the content and update it if necessary to stimulate the discussion.

@theresalech -san
When we receive the information from Joel-san, we would like to refer to it and proceed.
In that case, we want this Proposal to be a withdrawal.

**Comment by joeljfischer** [comment]
@Shohei-Kawano I did link it in the post above, here is the link again: https://github.com/smartdevicelink/sdl_evolution/pull/922

**Comment by Shohei-Kawano** [comment]
@joeljfischer -san
Oh.. sorry.. I didn't notice it was linked..
We will confirm it. Thank you.

**Comment by theresalech** [comment]
This proposal has been withdrawn per the author's request.  The author will investigate PR #922 as an alternate solution for the problem this proposal was looking to solve.
